rmw_opendds_cpp/src/rmw_client.cpp

```diff
@@ -12,259 +12,158 @@
-  } catch (const std::string& e) {
-    RMW_SET_ERROR_MSG(e.c_str());
+  } catch (const std::exception& e) {
+    RMW_SET_ERROR_MSG(e.what());
   } catch (...) {
     RMW_SET_ERROR_MSG("rmw_create_client failed");
   }
-  clean_client(*node_info, client);
-  if (buf) {
-    rmw_free(buf);
-  }
+  clean_client(client);
   return nullptr;
 }
 
 rmw_ret_t
 rmw_destroy_client(rmw_node_t * node, rmw_client_t * client)
 {
-  RMW_CHECK_FOR_NULL_WITH_MSG(node, "node is null", return RMW_RET_ERROR);
-  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(node, node->implementation_identifier,
-    opendds_identifier, return RMW_RET_ERROR)
-  auto node_info = static_cast<OpenDDSNodeInfo *>(node->data);
-  RMW_CHECK_FOR_NULL_WITH_MSG(node_info, "node_info is null", return RMW_RET_ERROR);
+  auto dds_client = DDSClient::from(client);
+  if (dds_client) {
+    if (dds_client->remove_from(OpenDDSNode::from(node))) {
+      clean_client(client);
+      return RMW_RET_OK;
+    }
+  }
+  return RMW_RET_ERROR;
+}
 
-  RMW_CHECK_FOR_NULL_WITH_MSG(client, "client is null", return RMW_RET_ERROR);
-  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(client, client->implementation_identifier,
-    opendds_identifier, return RMW_RET_ERROR)
+rmw_ret_t
+rmw_service_server_is_available(
+  const rmw_node_t * node,
+  const rmw_client_t * client,
+  bool * is_available)
+{
+  auto dds_node = OpenDDSNode::from(node);
+  if (!dds_node) {
+    return RMW_RET_ERROR;
+  }
+  auto dds_client = DDSClient::from(client);
+  if (!dds_client) {
+    return RMW_RET_ERROR;
+  }
+  //?? cmp DomainParticipant?
+  if (!is_available) {
+    RMW_SET_ERROR_MSG("is_available is null");
+    return RMW_RET_ERROR;
+  }
+  *is_available = dds_client->is_server_available();
+  return RMW_RET_OK;
+}
 
-  return clean_client(*node_info, client);
+rmw_ret_t
+rmw_send_request(
+  const rmw_client_t * client,
+  const void * ros_request,
+  int64_t * sequence_id)
+{
+  auto dds_client = DDSClient::from(client);
+  if (!dds_client) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (!ros_request) {
+    RMW_SET_ERROR_MSG("ros_request is null");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (!sequence_id) {
+    RMW_SET_ERROR_MSG("sequence_id is null");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  *sequence_id = dds_client->send(ros_request);
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+rmw_take_response(
+  const rmw_client_t * client,
+  rmw_service_info_t * request_header,
+  void * ros_response,
+  bool * taken)
+{
+  auto dds_client = DDSClient::from(client);
+  if (!dds_client) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (!request_header) {
+    RMW_SET_ERROR_MSG("request_header is null");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (!ros_response) {
+    RMW_SET_ERROR_MSG("ros_response is null");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (!taken) {
+    RMW_SET_ERROR_MSG("taken is null");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  *taken = dds_client->take(request_header, ros_response);
+  return RMW_RET_OK;
 }
+
```

Comment:
```suggestion
```

Reply:
The beginning empty line is also removed to match this.